tools/docker/Dockerfile

```diff
@@ -0,0 +1,231 @@
+# syntax=docker/dockerfile:experimental
+# Copyright 2018-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may
+# not use this file except in compliance with the License. A copy of the
+# License is located at
+#
+# 	http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+
+#########################################
+#
+# BUILD IMAGES
+#
+#########################################
+
+
+# Common tools needed for the build stages ahead. The final test images do not inherit directly from here, so this bloat
+# is dropped in those final end-use images.
+FROM debian:stretch as build-base
+ENV PATH /bin:/usr/bin:/usr/local/bin:/sbin:/usr/sbin:/usr/local/sbin:/usr/lib/go/bin
+RUN --mount=type=cache,id=aptcache,target=/var/cache/apt \
+	--mount=type=cache,id=aptlib,target=/var/lib/apt,sharing=locked \
+	mkdir -p /etc/apt/sources.list.d \
+	&& echo "deb http://ftp.debian.org/debian stretch-backports main" > /etc/apt/sources.list.d/stretch-backports.list \
+	&& apt update \
+	&& apt --target-release stretch-backports install -y golang-go \
+	&& apt install -y \
+		git \
+		wget \
+		build-essential \
+		libseccomp-dev \
+		btrfs-tools \
+		libdevmapper-dev \
+		musl-tools \
+		util-linux \
+		ca-certificates
+
+# Run as non-root for the rest of the build stages now that we've gotten apt installs out of the way.
+RUN useradd --create-home builder \
+	&& mkdir /output \
+	&& chown builder /output \
+	&& mkdir -p /home/builder/go
+
+# Buildkit caches don't support anything like a "--chown" flag yet, so ensure builder will have access to them
+RUN --mount=id=gomod,target=/home/builder/go/pkg/mod,type=cache \
+	chown -R builder /home/builder/go
```

Comment:
There's no guarantee that every `builder` user in each of these builds is going to have the same UID, meaning we could have some strange behavior when running parallel builds.  You can specify the UID in the `useradd` command above with the `--uid` flag, and that should make this safer.

(There are still issues with concurrent mutation of the Go module cache though, and the only way I can think of fixing that is to use a separate cache per build agent, but that negates some of the value of having a cache...)

Reply:
I just realized I forgot to address the second part of your comment here. I will make another quick PR update that uses the `sharing=locked` flag to prepopulate the cache in each stage and then in a second RUN step actually does the build with `sharing=private`, which should result in safe concurrent usage without locking the cache for too long.

Reply:
Got this working, but worth noting that once we upgrade to Go 1.12 we should be able to remove this workaround: https://github.com/golang/go/issues/26794